### ISSUE TITLE
Valid `.tar` archive inside `.gzip` archive not being extracted

### ISSUE THREAD

**Issue by rgmz** [trufflesecurity/trufflehog#2928]
Please review the [Community Note](https://github.com/trufflesecurity/trufflehog/blob/main/.github/community_note.md) before submitting

### TruffleHog Version

https://github.com/trufflesecurity/trufflehog/commit/b0fd70c0ffb7dd9e4dc2fb3e844f7fdadf332e9e

### Trace Output

 ```sh
$ trufflehog filesystem licenses.db
🐷🔑🐷  TruffleHog. Unearth your secrets. 🐷🔑🐷

2024-06-05T22:37:51-04:00       info-0  trufflehog      running source  {"source_manager_worker_id": "H6Lcl", "with_units": true}
2024-06-05T22:37:51-04:00       error   trufflehog      error unarchiving chunk.        {"source_manager_worker_id": "H6Lcl", "unit": "licenses.db", "unit_kind": "unit", "timeout": 30, "error": "error extracting archive with format: .tar: handling file: X11.txt: error creating custom reader: error identifying archive: matching 7z: zlib: invalid header"}
2024-06-05T22:37:51-04:00       info-0  trufflehog      finished scanning       {"chunks": 1229, "bytes": 13177071, "verified_secrets": 0, "unverified_secrets": 0, "scan_duration": "198.796129ms", "trufflehog_version": "dev"}
```

Actual trace with custom logging:
https://gist.github.com/rgmz/ec4cf437744a1f55132ed1830f0c4f4b

### Expected Behavior
The nested tar should extract.

### Actual Behavior

A valid tar inside a gzip is not extracted:
> 2024-06-05T22:37:51-04:00       error   trufflehog      error unarchiving chunk.        {"source_manager_worker_id": "H6Lcl", "unit": "licenses.db", "unit_kind": "unit", "timeout": 30, "error": "error extracting archive with format: .tar: handling file: X11.txt: error creating custom reader: error identifying archive: matching 7z: zlib: invalid header"}

**Works manually**
```sh
$ file -i licenses.db
licenses.db: application/gzip; charset=binary
$ gunzip licenses.db
$ ls
licenses
$ file -i licenses
licenses: application/x-tar; charset=binary
$ tar xf licenses
0BSD.hash               BCL.hash                       CC-BY-NC-SA-2.5.hash ...
```
### Steps to Reproduce

 1. Download https://github.com/kubernetes/git-sync/blob/b161f3f0c78b56f27188b4e4aabf672ba0b03706/vendor/github.com/google/licenseclassifier/licenses/licenses.db
 2. Run trufflehog and observe the reported error

## Environment
N/A

## Additional Context
N/A

### References
N/A

**Comment by rgmz** [comment]
It seems like the file `X11.txt` is being detected as an archive, for some reason. It's just a text file.

```
2024-06-07T14:41:23-04:00       info-0  trufflehog      Handling extracted file.        {"source_manager_worker_id": "I7eOu", "unit": "/tmp/licenses.db", "unit_kind": "unit", "timeout": 30, "filename": "WTFPL.txt", "size": 415}
2024-06-07T14:41:23-04:00       info-0  trufflehog      openArchive     {"source_manager_worker_id": "I7eOu", "unit": "/tmp/licenses.db", "unit_kind": "unit", "timeout": 30, "filename": "WTFPL.txt", "size": 415, "depth": 2}
2024-06-07T14:41:23-04:00       info-0  trufflehog      Handling extracted file.        {"source_manager_worker_id": "I7eOu", "unit": "/tmp/licenses.db", "unit_kind": "unit", "timeout": 30, "filename": "WTFPL.hash", "size": 2568}
2024-06-07T14:41:23-04:00       info-0  trufflehog      openArchive     {"source_manager_worker_id": "I7eOu", "unit": "/tmp/licenses.db", "unit_kind": "unit", "timeout": 30, "filename": "WTFPL.hash", "size": 2568, "depth": 2}
2024-06-07T14:41:23-04:00       info-0  trufflehog      Handling extracted file.        {"source_manager_worker_id": "I7eOu", "unit": "/tmp/licenses.db", "unit_kind": "unit", "timeout": 30, "filename": "X11.txt", "size": 1292}
2024-06-07T14:41:23-04:00       info-0  trufflehog      archive.extractorHandler: error creating custom reader  {"source_manager_worker_id": "I7eOu", "unit": "/tmp/licenses.db", "unit_kind": "unit", "timeout": 30, "filename": "X11.txt", "size": 1292, "error": "error identifying archive: matching rar: zlib: invalid header"}
error extracting archive with format: .tar: handling file: X11.txt: error creating custom reader: error identifying archive: matching rar: zlib: invalid header
panic: handling file: X11.txt: error creating custom reader: error identifying archive: matching rar: zlib: invalid header
```

**Comment by rgmz** [comment]
This _appears_ to be an issue with `mholt/archiver`, assuming that going archive.Decompressor > archive.Extractor is the intended use.

**Code**
```go
func main() {
    filename := "licenses.db"
    input, ferr := os.Open("/tmp/" + filename)
    if ferr != nil {
	    fmt.Printf("failed to open file: ")
	    panic(ferr)
    }
    
    //rdr := input
    rdr, rErr := readers.NewBufferedFileReader(input)
    if rErr != nil {
	    fmt.Printf("failed to create file reader")
	    panic(rErr)
    }
    run2(filename, rdr)
    return
}

// really polished code
func run2(filename string, rdr io.Reader) io.Reader {
	format, _, ferr := archiver.Identify(filename, rdr)
	if ferr != nil {
		fmt.Printf("failed to identify file")
		panic(ferr)
	}
	fmt.Printf("ARchive foramt is: %v\n", format.Name())

	switch archive := format.(type) {
	case archiver.Decompressor:
		fmt.Printf("Archive is: Decompressor\n")

		compReader, err := archive.OpenReader(rdr)
		if err != nil {
			panic(fmt.Errorf("error opening decompressor with format: %s %w", format.Name(), err))
		}

		run2(filename, compReader)
	case archiver.Extractor:
		fmt.Printf("Archive is: Extractor\n")

		handler := func(ctx con.Context, f archiver.File) error {
			fmt.Printf("Handling file: %s\n", f.Name())
			return nil
		}

		eErr := archive.Extract(context.Background(), rdr, nil, handler)
		fmt.Printf("archive.Extract error: %v\n", eErr)
		return nil
	default:
		fmt.Printf("Archive is unknown: %v\n", archive.Name())
		return nil
	}

	return nil
}
```

**Logs**

> Archive format is: .tar.gz
> Archive is: Decompressor
> ARchive foramt is: .tar
> Archive is: Extractor
> archive.Extract error: archive/tar: invalid tar header

**Comment by rgmz** [comment]
With that said, running this code against the `.tar` archive works without issue. There's something about the handling of `.gz` and then `.tar`.

**Comment by rgmz** [comment]
>  assuming that going archive.Decompressor > archive.Extractor is the intended use.

The `.tar.gz` file is of type [`archiver.CompressedArchive`](https://github.com/mholt/archiver/blob/de08bfa4c5588130c63dfcede2da42f26abccbb4/formats.go#L164). It seems that `CompressedArchive` is _both_ `archiver.Decompressor` and `archiver.Extractor`

This means that it matches both cases in the switch, but defaults to `Decompressor` because it's first in the list.

**Works**
```go
switch archive := format.(type) {
case archiver.Extractor:
  ...
case archiver.Decompressor:
  ...
default:
  ...
}
```

**Does not work**
```go
switch archive := format.(type) {
case archiver.Decompressor:
  ...
case archiver.Extractor:
  ...
default:
  ...
}
```

**Comment by rgmz** [comment]
Testing based on #2943, this error seems to occur when:

1. `archive.HandleFile` is called on `licenses.db` https://github.com/trufflesecurity/trufflehog/blob/440398815128e1f066ba5b49e65b1c0c3ecae200/pkg/handlers/archive.go#L64-L66
2. `archive.openArchive` calls `extractorHandler` https://github.com/trufflesecurity/trufflehog/blob/440398815128e1f066ba5b49e65b1c0c3ecae200/pkg/handlers/archive.go#L113-L114
3. `extractorHandler` calls `newFileReader` for the file `X11.txt` https://github.com/trufflesecurity/trufflehog/blob/440398815128e1f066ba5b49e65b1c0c3ecae200/pkg/handlers/archive.go#L186-L187
4. `newFileReader` calls `archiver.Identify` https://github.com/trufflesecurity/trufflehog/blob/440398815128e1f066ba5b49e65b1c0c3ecae200/pkg/handlers/handlers.go#L67-L68
5. [`archiver.Identify` returns an error other than `archiver.ErrNoMatch`](https://github.com/mholt/archiver/blob/de08bfa4c5588130c63dfcede2da42f26abccbb4/formats.go#L38), for some reason, which triggers the default case https://github.com/trufflesecurity/trufflehog/blob/440398815128e1f066ba5b49e65b1c0c3ecae200/pkg/handlers/handlers.go#L81-L82

**Comment by rgmz** [comment]
Fixed by #2959.
